### PR TITLE
#5187 Move all framework Shortcode Providers to the same namespace.

### PR DESCRIPTION
**Pull request by Firesphere**

### CODE REVIEW

**Comment by tractorcow** [comment]
This code looks fine to me!

File: Assets/File.php [hunk 1]
```diff
@@ -234,92 +234,6 @@ public static function get_shortcodes() {
...
-	 * Replace "[file_link id=n]" shortcode with an anchor tag or link to the file.
-	 *
-	 * @param array $arguments Arguments passed to the parser
-	 * @param string $content Raw shortcode
-	 * @param ShortcodeParser $parser Parser
-	 * @param string $shortcode Name of shortcode used to register this handler
-	 * @param array $extra Extra arguments
-	 * @return string Result of the handled shortcode
-	 */
-	public static function handle_shortcode($arguments, $content, $parser, $shortcode, $extra = array()) {
```
Comment (tractorcow) [inline Assets/File.php]: Can you please add this to the 4.0.0 changelog under the appropriate section? (Probably under Filesystem API Additions / Changes).

File: Forms/HTMLEditor/FileShortcodeProvider.php [hunk 2]
```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace SilverStripe\Forms\HtmlEditor;
```
Comment (sminnee) [inline Forms/HTMLEditor/FileShortcodeProvider.php]: I don't think this is the right namespace – this shortcode handler should be part of the `SilverStripe\Assets` subsystem. If you didn't have `SilverStripe\Assets` you wouldn't have this shortcode provider available. This would be like putting the models of every SilverStripe module into the `SilverStripe\ORM` namespace.

Perhaps this can go into the `SilverStripe\Assets\ViewSupport` namespace, since it's a support class to assist the view?

Obviously you could argue: "if you want to use silverstripe\assets without the silverstripe WYSIWYG editor, then you've got the same problem in reverse". However, I think that we're more likely to require the WYSIWYG editor without assets than the reverse.

File: Assets/File.php (hunk 1 above)
Comment (Firesphere) [inline Assets/File.php]: Yep, will do.

File: Forms/HTMLEditor/FileShortcodeProvider.php (hunk 2 above)
Comment (Firesphere) [inline Forms/HTMLEditor/FileShortcodeProvider.php]: I went with this namespace, as the previous extraction, the EmbedShortcodeProvider, was decided to go into this namespace.

I'm happy to change it, but I'd like to have them all bundled?

My 2 cents on the namespace: `SilverStripe\View\Provider` or `SilverStripe\View\Shortcode` (Or something among those lines). as they are part of the View, not Forms or Assets as such.

**Comment by Firesphere** [comment]
I'll be adding commits, so it's easier to track the changes, until it's a go, then squash it. That okay with you folks?

File: Forms/HTMLEditor/FileShortcodeProvider.php (hunk 2 above)
Comment (Firesphere) [inline Forms/HTMLEditor/FileShortcodeProvider.php]: Update on previous comment:
Wouldn't it be more logic to put all the Shortcode Parsers under the same namespace? As the interface is defined under `SilverStripe\View\Parsers`, I would suggest moving all providers to the same namespace?

File: Assets/File.php (hunk 1 above)
Comment (Firesphere) [inline Assets/File.php]: Added changes to the changelog. Including instructions on how to migrate.

File: Forms/HTMLEditor/FileShortcodeProvider.php (hunk 2 above)
Comment (Firesphere) [inline Forms/HTMLEditor/FileShortcodeProvider.php]: I feel slightly uncomfortable, moving the providers to separate namespaces, as the embed provider would stay in the Form\HtmlEditor?
I've for now moved them all into `SilverStripe\View\Parsers`. Welcome any thoughts? Happy to move them around, but I'd personally prefer to keep all providers in the same namespace, so CMS Providers (like sitetree_link), can go into `SilverStripe\CMS\Chosen\Namespace` as well, to keep it all in a unified place.

**Comment by tractorcow** [comment]
@sminnee I don't mind the new namespace. Is `SilverStripe\View\Parsers` ok with you? If so I'll merge.

**Comment by Firesphere** [comment]
Have rebased and squashed into a single commit. Separately, the tests are fine. Unsure of anything I rebased on broke other tests though.

**Comment by Firesphere** [comment]
@tractorcow It's not a "new" namespace. It's the same namespace as where the ShortcodeParser Interface resides.

**Comment by sminnee** [comment]
My concern here is that this makes the assets module a dependency of the view module, which doesn't make sense to me.

**Comment by Firesphere** [comment]
tl;dr: It makes OembedProvider a bit a hovering orphan that needs a good namespace.

I understand and also partially agree with your concern @sminnee.

What I however wish to accomplish, is unity. Hence also the move of the Oembed provider to the same namespace.

If we put the File and Image providers in Assets\ViewSupport, I think the Oembed one, should also move to a similar namespace instead of HTMLEditor\Form.

I'm happy with the ViewSupport namespace, but unclear on how to make it consistent across all providers (same goes for CMS, which would probably then go to CMS\Model\ViewSupport?)

**Comment by dhensby** [comment]
I'm a bit split on this but I'm learning on agreement with sam here. We should be placing parsers in namespaces relevant to the parser, not lumping them together in one.

**Comment by Firesphere** [comment]
I can live with that @dhensby, I'm trying to find a good concensus, where the View was my first thought on unifying.

I'm just still stuck on where the Oembed provider should go. As having a single `Oembed\ViewSupport` namespace for it, sounds silly to me.

**Comment by sminnee** [comment]
You can leave the Oembed provider where it is if you want.

**Comment by Firesphere** [comment]
To unify, Do you mind if I make the Oembed `SilverStripe\View\ViewSupport` and create a folder for it like that?
I want the end namespace to be as uniform as possible, to avoid confusion.

**Comment by dhensby** [comment]
@Firesphere just do it :) Let's get this merged

**Comment by Firesphere** [comment]
Agreed and fixed your change request @sminnee

**Comment by dhensby** [comment]
needs rebasing

**Comment by Firesphere** [comment]
Will do when I have time tomorrow or this weekend :)

**Comment by Firesphere** [comment]
I completely lost sight of this PR, have rebased but it seems there are still merge conflicts?

**Comment by chillu** [comment]
@Firesphere Any progress here?

**Comment by unclecheese** [comment]
Closing. Reopened as:
* https://github.com/silverstripe/silverstripe-assets/pull/40
* https://github.com/silverstripe/silverstripe-framework/pull/7085
